Sources/SwiftFusion/Inference/PenguinExtensions.swift

```diff
@@ -104,6 +137,577 @@ extension ArrayBuffer: CustomDebugStringConvertible {
+
+  public func _customIndexOfEquatableElement(_ element: Element) -> Index?? {
+    vtable[0].customIndexOfEquatableElement(self, element)
+  }
+
+  public func _customLastIndexOfEquatableElement(_ element: Element) -> Index?? {
+    vtable[0].customLastIndexOfEquatableElement(self, element)
+  }
+
+  /// Returns an index that is the specified distance from the given index.
+  public func index(_ i: Index, offsetBy distance: Int) -> Index {
+    vtable[0].indexOffsetBy(self, i, distance)
+  }
+
+  /// Returns an index that is the specified distance from the given index,
+  /// unless that distance is beyond a given limiting index.
+  public func index(
+    _ i: Index, offsetBy distance: Int, limitedBy limit: Index
+  ) -> Index? {
+    vtable[0].indexOffsetByLimitedBy(self, i, distance, limit)
+  }
+
+  /// Returns the distance between two indices.
+  public func distance(from start: Index, to end: Index) -> Int {
+    vtable[0].distance(self, start, end)
+  }
+
+  public mutating func _withUnsafeMutableBufferPointerIfSupported<R>(
+    _ body: (inout UnsafeMutableBufferPointer<Element>) throws -> R
+  ) rethrows -> R? {
+    var result: R? = nil
+    try vtable[0].existentialDispatch.withUnsafeMutableBufferPointerIfSupported(&storage) {
+      var buf = UnsafeMutableBufferPointer(
+        start: $0?.assumingMemoryBound(to: Element.self), count: $1)
+      try result = body(&buf)
+    }
+    return result
+  }
+
+  public mutating func withContiguousMutableStorageIfAvailable<R>(
+    _ body: (inout UnsafeMutableBufferPointer<Element>) throws -> R
+  ) rethrows -> R? {
+    var result: R? = nil
+    try vtable[0].existentialDispatch.withContiguousMutableStorageIfAvailable(&storage) {
+      var buf = UnsafeMutableBufferPointer(
+        start: $0?.assumingMemoryBound(to: Element.self), count: $1)
+      try result = body(&buf)
+    }
+    return result
+  }
+  
+  /// Reorders the elements of the collection such that all the elements
+  /// that match the given predicate are after all the elements that don't
+  /// match.
+  ///
+  /// - Complexity: O(*n*), where *n* is the length of the collection.
+  public mutating func partition(
+    by belongsInSecondPartition: (Element) throws -> Bool
+  ) rethrows -> Index {
+    var r: Index? = nil
+    try vtable[0].existentialDispatch.partition(&storage, &r) { p in
+      try belongsInSecondPartition(p.assumingMemoryBound(to: Element.self).pointee)
+    }
+    return r.unsafelyUnwrapped
+  }
+
+  /// Exchanges the values at the specified indices of the collection.
+  public mutating func swapAt(_ i: Index, _ j: Index) {
+    vtable[0].swapAt(&self, i, j)
+  }
+}
+
+// =================================================================================================
+
+public struct FlattenedMutableCollection<Base: MutableCollection>
+  where Base.Element: MutableCollection
+{
+  var base: Base
+    
+  init(_ base: Base) {
+    self.base = base
+  }
+}
+
```

Comment:
Could this somehow be used to implement `FlattenedScalars`?

If this used a closure `(Base.Element) -> InnerCollection` to extract the inner collection then `FlattenedScalars` could be implemented as a `FlattenedMutableCollection` with `\.scalars` for the closure. Though I guess that probably wouldn't optimize well.

Reply:
It certainly *might* not optimize well, and it's not a chance I wanted to take.